overlord/state/warning.go

```diff
@@ -0,0 +1,245 @@
+	var jw jsonWarning
+	err := json.Unmarshal(data, &jw)
+	if err != nil {
+		return err
+	}
+	w.message = jw.Message
+	w.firstSeen = jw.FirstSeen
+	w.lastSeen = jw.LastSeen
+	if jw.LastShown != nil {
+		w.lastShown = *jw.LastShown
+	}
+	w.deleteAfter, err = time.ParseDuration(jw.DeleteAfter)
+	if err != nil {
+		return err
+	}
+	w.repeatAfter, err = time.ParseDuration(jw.RepeatAfter)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *Warning) IsDeletable(now time.Time) bool {
+	return w.lastSeen.Add(w.deleteAfter).Before(now)
+}
+
+func (w *Warning) IsShowable(t time.Time) bool {
+	return (w.lastShown.IsZero() || w.lastShown.Add(w.repeatAfter).Before(t)) && !w.firstSeen.After(t)
+}
+
+// flattenWarning loops over the warnings map, and returns all
+// warnings therein as a flat list, for serialising.
+// Call with the lock held.
+func (s *State) flattenWarnings() []*Warning {
+	flat := make([]*Warning, 0, len(s.warnings))
+	for _, w := range s.warnings {
+		flat = append(flat, w)
+	}
+	return flat
+}
+
+// unflattenWarnings takes a flat list of warnings and replaces the
+// warning map with them.
+// Call with the lock held.
+func (s *State) unflattenWarnings(flat []*Warning) {
+	s.warnings = make(map[string]*Warning, len(flat))
+	for _, w := range flat {
+		s.warnings[w.message] = w
+	}
+}
+
+// AddWarning records a warning: if it's the first Warning with this
+// message it'll be added (with its firstSeen and lastSeen set to the
+// current time), otherwise the existing one will have its lastSeen
+// updated.
+func (s *State) AddWarning(message string) {
+	s.addWarningFull(Warning{
+		message:     message,
+		deleteAfter: DefaultDeleteAfter,
+		repeatAfter: DefaultRepeatAfter,
+	}, time.Now().UTC())
+}
+
+func (s *State) addWarningFull(w Warning, t time.Time) {
+	s.writing()
+
+	if s.warnings[w.message] == nil {
+		w.firstSeen = t
+		s.warnings[w.message] = &w
+	}
+	s.warnings[w.message].lastSeen = t
+}
+
+// DeleteOldWarnings delets warnings that are up for deletion.
+func (s *State) DeleteOldWarnings() int {
+	s.writing()
+
+	now := time.Now().UTC()
+
+	n := 0
+	for k, w := range s.warnings {
+		if w.IsDeletable(now) {
+			delete(s.warnings, k)
+			n++
+		}
+	}
+	return n
+}
+
+type byLastSeen []*Warning
+
+func (a byLastSeen) Len() int           { return len(a) }
+func (a byLastSeen) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byLastSeen) Less(i, j int) bool { return a[i].lastSeen.Before(a[j].lastSeen) }
+
+// AllWarnings returns all the warnings in the system, whether they're
+// due to be shown or not. They'll be sorted by lastSeen.
+func (s *State) AllWarnings() []*Warning {
+	s.reading()
+
+	all := s.flattenWarnings()
+	sort.Sort(byLastSeen(all))
+
+	return all
+}
+
+// OkayWarnings marks warnings that were showable at the given time as shown.
+func (s *State) OkayWarnings(t time.Time) int {
+	t = t.UTC()
+	s.writing()
+
+	n := 0
+	for _, w := range s.warnings {
+		if w.IsShowable(t) {
```

Comment:
IsShowable uses/checks firstSeen, but here at Okaying should we use lastSeen perhaps? should we okay or not warnings that were emitting again since t?

Reply:
translating to the new names (just so I don't get confused),
> `IsShowable` uses/checks `firstAdded`, but here at Okaying should we use `lastAdded` perhaps? should we okay or not warnings that were emitting again since `t`?

A warning is showable at time `t` iff

* it was known at `t` (that is, `firstAdded` is not after `t`)
* it had not been shown at `t` (that is, `lastShown` is zero), or it had been shown long before `t` (`repeatAfter` before).

if a warning that would be showable at `t` is added again at `tʹ > t`, then `IsShowable(t)` will be `true`, and its `lastShown` will be updated, but it will now have `lastAdded == tʹ` so it will be again showable after `repeatAfter` has elapsed.

This is, as I understand it, part of the design: if the system is throwing up a warning periodically, `snap okay` should shut the warning up for a while (until `repeatAfter`).